### ISSUE TITLE
trial -u can fail doctests on second pass

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2713]
|<img alt="mgeisler's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| mgeisler reported|
|-|-|
|Trac ID|trac#2713|
|Type|defect|
|Created|2007-06-22 21:50:28Z|
|Branch|https://github.com/twisted/twisted/tree/rerundoctest-2713|

Given a file `a.py` containing a doctest like this:

```
def foo():
    """
    >>> foo()
    123
    """
    return 123
```

This doctest passes when used once with trial, but fails if trial is run with the --until-failure flag:

```
% trial -u trial_doctest_bug
Test Pass 1         
Running 1 tests.
doctest
  DocTestCase
    foo ...                                                            [OK]

---------------------------------------------------------------------------
Ran 1 tests in 0.030s

PASSED (successes=1)
Test Pass 2
Running 1 tests.
doctest
  DocTestCase
    foo ...                                                          [FAIL]

===========================================================================
[FAIL]: Doctest: trial_doctest_bug.a.foo

Traceback (most recent call last):
  File "unittest.py", line 260, in run
    testMethod()
  File "doctest.py", line 2157, in runTest
    raise self.failureException(self.format_failure(new.getvalue()))
exceptions.AssertionError: Failed doctest test for trial_doctest_bug.a.foo
  File "/tmp/trial_doctest_bug/a.py", line 2, in foo

----------------------------------------------------------------------
File "/tmp/trial_doctest_bug/a.py", line 5, in trial_doctest_bug.a.foo
Failed example:
    foo()
Exception raised:
    Traceback (most recent call last):
      File "doctest.py", line 1248, in __run
        compileflags, 1) in test.globs
      File "<doctest trial_doctest_bug.a.foo[0]>", line 1, in ?
        foo()
    NameError: name 'foo' is not defined

---------------------------------------------------------------------------
Ran 1 tests in 0.005s

FAILED (failures=1)
```

If I change the doctest so that it explicitly imports everything:

```
def foo():
    """
    >>> from trial_doctest_bug.a import foo
    >>> foo()
    123
    """
    return 123
```

then the doctest works when run repeatedly by trial.

Is this the indented behaviour?

Attachments:

* [DocTestCase-issue2604-workaround.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/189/189dfab3335a729ceec7f86495df5620235b5882/9fa51445a210d22d1ba7e27932d76910dca0bf77.patch) (2047 bytes) - added by pjd on 2008-04-16 12:04:57Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2713 2713
type__defect defect
reporter__mgeisler mgeisler
priority__normal normal
milestone__ 
branch__branches_rerundoctest_2713 branches/rerundoctest-2713
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1182549028000000 1182549028000000
changetime__1212496203000000 1212496203000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="mgeisler's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|mgeisler commented|
|-|-|

I forgot to mention that I see this behavior with a fresh SVN checkout (revision 20548).

**Comment by twisted-trac** [comment]
|<img alt="pjd's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|pjd set owner to pjd<br>pjd set status to `assigned`|
|-|-|

This is a regression in `doctest.DocTestCase` in Python 2.4+;  see [issue2604](http://bugs.python.org/issue2604).

**Comment by twisted-trac** [comment]
|<img alt="pjd's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|pjd set owner to @exarkun<br>pjd set status to `new`|
|-|-|

The attached patch adds a workaround and test for the problem.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [23672]) Branching to 'rerundoctest-2713'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [23683]) Merge rerundoctest-2713

Author: pjd
Reviewer: exarkun
Fixes: [#2713](https://github.com/twisted/twisted/issues/2713)

Work-around the behavior of DocTestSuite in Python 2.4 and newer which
causes doctests to lose their globals on second and subsequent runs (eg
during --until-failure runs) by saving and then restoring those globals.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `reopened`|
|-|-|

(In [23684]) Revert r23682 - Python 2.3 doctest regression

Reopens: [#2713](https://github.com/twisted/twisted/issues/2713)

Python 2.3 builder has a total failure inducing problem:

```
Traceback (most recent call last):
  File "./bin/trial", line 24, in ?
    run()
  File "/srv/bb-slave/Run/slave/full2.3/Twisted/twisted/scripts/trial.py", line 357, in run
    suite = _getSuite(config)
  File "/srv/bb-slave/Run/slave/full2.3/Twisted/twisted/scripts/trial.py", line 315, in _getSuite
    return loader.loadByNames(config['tests'], recurse)
  File "/srv/bb-slave/Run/slave/full2.3/Twisted/twisted/trial/runner.py", line 638, in loadByNames
    suites = [self.loadAnything(thing, recurse)
  File "/srv/bb-slave/Run/slave/full2.3/Twisted/twisted/trial/runner.py", line 596, in loadAnything
    return self.loadPackage(thing, recurse)
  File "/srv/bb-slave/Run/slave/full2.3/Twisted/twisted/trial/runner.py", line 551, in loadPackage
    thingToAdd = self.loadModule(module)
  File "/srv/bb-slave/Run/slave/full2.3/Twisted/twisted/trial/runner.py", line 475, in loadModule
    docSuite.addTest(self.loadDoctests(doctest))
  File "/srv/bb-slave/Run/slave/full2.3/Twisted/twisted/trial/runner.py", line 581, in loadDoctests
    return doctest.DocTestSuite(module, setUp=saveGlobs)
TypeError: DocTestSuite() got an unexpected keyword argument 'setUp'
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Someone else should take a look at this now I suppose.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve removed owner|
|-|-|

Looks fine, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [23851]) Merge rerundoctest-2713

Author: pjd, exarkun
Reviewer: therve
Fixes: [#2713](https://github.com/twisted/twisted/issues/2713)

Work around a regression in doctest in Python 2.4 and newer which causes the
globals for a test to be discarded after the first run.  This prevents most
doctests from ever succeeding more than once when run with trial's
--until-failure option.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
